File: problem_059.rs
// -*- compile-command: "rustc -o problem_022_rs problem_022.rs" -*-
// Copyright (c) 2016 Michael Caldwell
use std::fs::File;
use std::io::Read;

fn decipher(data: &Vec<&str>, key: [u8; 3]) -> u32 {
    let mut sum: u32 = 0;
    let mut spaces = 0;
    let mut index = 0;

    for c in data {
        // Perform XOR operation using KEY
        let new_c: u8 = c.parse::<usize>().unwrap() as u8 ^ key[index % 3];
        sum += new_c as u32;
        //print!("{}", new_c as char); // DEBUG - Print deciphered text

        // See if character is out of range, and reject cipher
        if new_c < ' ' as u8 || new_c > 'z' as u8 { return 0; }

        // Make sure no longer than 16 characters between spaces
        if new_c == ' ' as u8 {
            spaces = 0;
        } else {
            spaces += 1;
        }

        if spaces > 16 {
            return 0;
        }

        // Next char
        index += 1;
    }

    return sum;
}

fn main() {
    println!("Project Euler - Problem 59:");
    println!("Decrypt the message and find the sum of the ASCII values in the original text.\n");

    // Import data from file
    let mut data = String::new();
    let mut f = File::open("problem_059.txt").unwrap();
    f.read_to_string(&mut data).unwrap();

    // Split character codes into vector
    let chars: Vec<&str> = data.split(',').collect();

    // Iterate through all 3-letter keys
    for a in ('a' as u8)..('z' as u8) {
        for b in ('a' as u8)..('z' as u8) {
            for c in ('a' as u8)..('z' as u8) {
                let sum = decipher(&chars, [a, b, c]);
                if sum != 0 {
                    println!("{} (Key: {}{}{})", sum, a as char, b as char, c as char);
                    return;
                }
            }
        }
    }

    println!("No solution found.");
    return;
}

File: problem_059.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "xcrun -sdk macosx swiftc problem_059.swift -o problem_059_swift" -*-
// Copyright (c) 2016 Michael Caldwell
import Foundation

func decipher(a: Int, b: Int, c: Int, data: [String]) -> Int {
    let cipher = [a, b, c]
    var deciphered = ""
    var wordLength = 0
    var count = 0

    for i in 0 ..< data.count {
        // XOR and append to string
        let char = Int(data[i])! ^ (cipher[i%3] + Int(("a" as UnicodeScalar).value))
        count += char
        deciphered += String(format: "%c", char) as String

        // Look for unlikely characters
        if char > Int(("z" as UnicodeScalar).value) || char < Int((" " as UnicodeScalar).value) {
            return 0
        }

        // Look for unlikely word lengths
        wordLength += 1
        if char == Int((" " as UnicodeScalar).value) { wordLength = 0 }
        if wordLength > 13 { return 0 }
    }

    // If we get here, we found a probably valid message; Print it!
    let startingValue = Int(("a" as UnicodeScalar).value)
    print(String(format: "Cipher: %c%c%c", a+startingValue, b+startingValue, c+startingValue) as String)
    print(deciphered)
    print("")

    return count
}

print("Project Euler - Problem 59:")
print("Decrypt the message and find the sum of the ASCII values in the original text.\n")

// Import data from file
let file: FileHandle? = FileHandle(forReadingAtPath: "problem_059.txt")
if file == nil {
    print("Unable to open file.")
    exit(1)
}

// Read all the data
let data = file?.readDataToEndOfFile()

// Close the file
file?.closeFile()

// Convert our data to string
let dataAsString = String(data: data!, encoding: String.Encoding.utf8)

// Separate numbers into array
var array = dataAsString!.components(separatedBy:",")

// Brute force!
var result = 0

for A in 0...25 {
    for B in 0...25 {
        for C in 0...25 {
            result = decipher(a:A, b:B, c:C, data:array)
            if result != 0 { break }
        }
        if result != 0 { break }
    }
    if result != 0 { break }
}

print(result)
